packages/backend/src/apps/m365-excel/common/oauth/token-cache.ts

```diff
@@ -0,0 +1,55 @@
+import { Mutex } from 'async-mutex'
+
+import {
+  isM365TenantKey,
+  M365TenantKey,
+  m365TenantKeys,
+} from '@/config/app-env-vars/m365'
+import { type IHttpClient } from '@/helpers/http-client'
+
+import {
+  makeAccessTokenRequest,
+  type MsGraphAccessToken,
+} from './token-request'
+
+interface CachedToken {
+  mutex: Mutex // Non-blocking async mutex
+  token: MsGraphAccessToken | null
+}
+
+const cachedTokens: Record<M365TenantKey, CachedToken> = Object.create(null)
+for (const key of m365TenantKeys) {
+  cachedTokens[key] = {
+    mutex: new Mutex(),
+    token: null,
+  }
+}
+
+export async function getAccessToken(
+  tenantKey: string,
+  httpClient: IHttpClient,
+): Promise<string> {
+  if (!isM365TenantKey(tenantKey)) {
+    throw new Error(`${tenantKey} is an invalid M365 tenant`)
+  }
+  const cachedToken = cachedTokens[tenantKey]
+
+  if (Date.now() < cachedToken.token?.expiryTimestamp) {
+    return cachedToken.token.value
+  }
+
+  // Slightly convoluted to avoid grabbing the mutex for every single request.
+  await cachedToken.mutex.runExclusive(async () => {
```

Comment:
can i just check that the mutex here is to ensure that multiple requests are executed one by one rather than simultaneously and check if the current token can still be used? im just thinking what happens without the mutex ps this one a bit dumb question

Reply:
Pretty much yes. 

If you want to be precise, the mutex only serializes acquisition of auth token from M365. Other parts of the request (e.g. writing to excel, etc) are allowed to run concurrently.

The goal is to prevent a case where we have like 1000 excel actions queued up, and all of them try to acquire an auth token at once (which causes rate limit error).

Reply:
this is pretty useful, especially if a worker had to restart after a long while. think it could be done with simple closures? [[gist](https://gist.github.com/pregnantboy/87ac439e0a8a755c8ea4e51b16de9b27)]